File: TxtFileExporter.java
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.time.LocalDate;
import java.util.Map;
import java.util.TreeMap;
public class TxtFileExporter implements IFileExporter {
    private String outputFileName;
    public TxtFileExporter(String outputFileName) {
        this.outputFileName = outputFileName;
    }
    public String getOutputFileName() {
        return outputFileName;
    }
    public void setOutputFileName(String outputFileName) {
        this.outputFileName = outputFileName;
    }
    @Override
    public void export(String habitName, Map<LocalDate, Boolean> logs) {
        try (PrintWriter writer = new PrintWriter(new FileWriter(outputFileName))) {
            writer.println("===========================================");
            writer.println("       HABIT TRACKER REPORT");
            writer.println("===========================================");
            writer.println("Habit: " + habitName);
            writer.println("Generated: " + LocalDate.now());
            writer.println("===========================================");
            writer.println();
            TreeMap<LocalDate, Boolean> sortedLogs = new TreeMap<>(logs);
            int totalDays = sortedLogs.size();
            int completedDays = 0;
            writer.println("Daily Log:");
            writer.println("-------------------------------------------");
            for (Map.Entry<LocalDate, Boolean> entry : sortedLogs.entrySet()) {
                String status = entry.getValue() ? "[X] Completed" : "[ ] Not Completed";
                writer.println(entry.getKey() + " - " + status);
                if (entry.getValue()) {
                    completedDays++;
                }
            }
            writer.println("-------------------------------------------");
            writer.println();
            writer.println("Summary:");
            writer.println("Total Days Tracked: " + totalDays);
            writer.println("Days Completed: " + completedDays);
            writer.println("Days Missed: " + (totalDays - completedDays));
            if (totalDays > 0) {
                double percentage = (completedDays * 100.0) / totalDays;
                writer.println("Completion Rate: " + String.format("%.1f", percentage) + "%");
            }
            writer.println("===========================================");
            System.out.println("Report exported successfully to " + outputFileName);
        } catch (IOException e) {
            System.err.println("Error exporting report: " + e.getMessage());
        }
    }
}

File: TxtFileExporter.py
from datetime import date
from typing import Dict
import csv

class TxtFileExporter:
    def __init__(self, output_file_name: str):
        self.output_file_name = output_file_name

    def get_output_file_name(self) -> str:
        return self.output_file_name

    def set_output_file_name(self, output_file_name: str) -> None:
        self.output_file_name = output_file_name

    def export(self, habit_name: str, logs: Dict[date, bool]) -> None:
        try:
            with open(self.output_file_name, 'w', newline='') as file:
                file.write("===========================================\n")
                file.write("       HABIT TRACKER REPORT\n")
                file.write("===========================================\n")
                file.write(f"Habit: {habit_name}\n")
                file.write(f"Generated: {date.today()}\n")
                file.write("===========================================\n")
                file.write("\n")
                sorted_logs = dict(sorted(logs.items()))
                total_days = len(sorted_logs)
                completed_days = 0
                file.write("Daily Log:\n")
                file.write("-------------------------------------------\n")
                for log_date, status in sorted_logs.items():
                    status_str = "[X] Completed" if status else "[ ] Not Completed"
                    file.write(f"{log_date} - {status_str}\n")
                    if status:
                        completed_days += 1
                file.write("-------------------------------------------\n")
                file.write("\n")
                file.write("Summary:\n")
                file.write(f"Total Days Tracked: {total_days}\n")
                file.write(f"Days Completed: {completed_days}\n")
                file.write(f"Days Missed: {total_days - completed_days}\n")
                if total_days > 0:
                    percentage = (completed_days * 100.0) / total_days
                    file.write(f"Completion Rate: {percentage:.1f}%\n")
                file.write("===========================================\n")
                print(f"Report exported successfully to {self.output_file_name}")
        except Exception as e:
            print(f"Error exporting report: {str(e)}")